src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -0,0 +1,3657 @@
+    # Copy the remaining unique keys from new
+    for key in newkeys:
+        merged[key] = copy.deepcopy(new[key])
+    return merged
+
+
+# Convert condensed list format (with ranges) to a Python list
+def expand_list(input):
+    # The input string is a comma separated list of digits and ranges.
+    # Examples include:
+    # 0-3,8-11
+    # 0,2,4,6
+    # 2,5-7,10
+    list = []
+    s = input.strip()
+    if not s:
+        return list
+    for r in s.split(','):
+        if '-' in r[1:]:
+            start, end = r.split('-', 1)
+            for i in range(int(start), int(end) + 1):
+                list.append(i)
+        else:
+            list.append(int(r))
+    return list
+
+
+# Return a list of files similar to the find command
+def find_files(path, name=None, kind=None):
+    result = []
+    for root, dirs, files in os.walk(path):
+        if name is None:
+            filtered = files
+        else:
+            filtered = fnmatch.filter(files, name)
+        for f in filtered:
+            p = os.path.join(root, f)
+            if kind is None:
+                result.append(p)
+                continue
+            m = os.lstat(p).st_mode
+            if 'f' in kind and stat.S_ISREG(m):
+                result.append(p)
+            elif 'l' in kind and stat.S_ISLNK(m):
+                result.append(p)
+            elif 'c' in kind and stat.S_ISCHR(m):
+                result.append(p)
+            elif 'b' in kind and stat.S_ISBLK(m):
+                result.append(p)
+            elif 'p' in kind and stat.S_ISFIFO(m):
+                result.append(p)
+            elif 's' in kind and stat.S_ISSOCK(m):
+                result.append(p)
+        if name is None:
+            filtered = dirs
+        else:
+            filtered = fnmatch.filter(dirs, name)
+        for d in filtered:
+            p = os.path.join(root, d)
+            if kind is None:
+                result.append(p)
+                continue
+            m = os.lstat(p).st_mode
+            if 'd' in kind and stat.S_ISDIR(m):
+                result.append(p)
+            elif 'l' in kind and stat.S_ISLNK(m):
+                result.append(p)
+    return result
+
+
+# Return a properly cast zero value
+def initialize_resource(resc):
+    if isinstance(resc, pbs.pbs_int):
+        return pbs.pbs_int(0)
+    if isinstance(resc, pbs.pbs_float):
+        return pbs.pbs_float(0)
+    if isinstance(resc, pbs.size):
+        return pbs.size('0')
+    if isinstance(resc, int):
+        return 0
+    if isinstance(resc, float):
+        return 0.0
+    if isinstance(resc, list):
+        return []
+    if isinstance(resc, dict):
+        return {}
+    if isinstance(resc, tuple):
+        return ()
+    if isinstance(resc, str):
+        return ''
+    return resc
+
+
+# Helper function to ignore suspended jobs when removing
+# "already used" resources
+def job_to_be_ignored(jobid):
+    # Get job substate from small utility that calls printjob
+    pbs_exec = ''
```

Comment:
We call printjob because it reads job data from the local file rather than contacting the server. Otherwise, the operation scales poorly.